src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -150,31 +133,164 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
         }
 
         /// <summary>
-        /// Reads an array of element type <typeparamref name="TAnyElement"/> from the given buffer <paramref name="buf"/>.
+        /// Reads a generic list containing elements of type <typeparamref name="TRequestedElement"/> from the given buffer <paramref name="buf"/>.
         /// </summary>
-        protected async ValueTask<List<TAnyElement>> ReadList<TAnyElement>(NpgsqlReadBuffer buf, bool async)
+        protected async ValueTask<List<TRequestedElement>> ReadList<TRequestedElement>(NpgsqlReadBuffer buf, bool async)
         {
             await buf.Ensure(12, async);
             var dimensions = buf.ReadInt32();
 
             if (dimensions == 0)
-                return new List<TAnyElement>();
+                return ElementTypeInfo<TRequestedElement>.EmptyList;
             if (dimensions > 1)
-                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(TAnyElement).Name}>");
+                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(TRequestedElement).Name}>");
+
+            var containsNulls = buf.ReadInt32() == 1;
+            if (!ElementTypeInfo<TRequestedElement>.IsReferenceOrNullable && containsNulls)
+                throw UnsupportedNonNullableTypeException($"List<{typeof(TRequestedElement).Name}>", $"List<{typeof(TRequestedElement).Name}?>");
 
-            buf.ReadInt32();  // Has nulls. Not populated by PG?
             buf.ReadUInt32(); // Element OID. Ignored.
 
             await buf.Ensure(8, async);
             var length = buf.ReadInt32();
             buf.ReadInt32(); // We don't care about the lower bounds
 
-            var list = new List<TAnyElement>(length);
+            var list = new List<TRequestedElement>(length);
             for (var i = 0; i < length; i++)
-                list.Add(await _elementHandler.ReadWithLength<TAnyElement>(buf, async));
+                list.Add(await ElementHandler.ReadWithLength<TRequestedElement>(buf, async));
             return list;
         }
 
+        static InvalidOperationException UnsupportedNonNullableTypeException(string requestedType, string suggestedType)
+            => new InvalidOperationException(
+                $"Can't read a non-nullable {requestedType} if the database array field contains null values. Use {suggestedType} instead.");
+
+        #endregion Read
+
+        #region Static generic caching helpers
+
+        internal static class ElementTypeInfo<TElement>
+        {
+            public static readonly bool IsReferenceOrNullable =
+                !typeof(TElement).IsValueType ||
+                typeof(TElement).IsGenericType && typeof(TElement).GetGenericTypeDefinition() == typeof(Nullable<>);
```

Comment:
You can invert this by renaming the field to `IsNonNullable` or `NotNullble` (as you wish) and changing expression to `typeof(TElement).IsValueType && Nullable.GetUnderlyingType(typeof(TElement)) is null`.

Reply:
Good suggestion!
I'll change it.